### DISCUSSION TITLE
Integrating within a HomeSeer network/installation?

### DISCUSSION THREAD

**Discussion by benze** [zwave-js/node-red-contrib-zwave-js#156] [Q&A]
I have a HomeSeer installation with a ZWave controller.  I would like to start using NodeRed for ZWave control, but can't figure out how to make both pieces work in parallel.

I know that HomeSeer has a NodeRed plugin that allows flow control of HomeSeer devices, but would like to attempt to control devices from NodeRed directly.  

Is this feasible?  Is the only option to include a secondary controller in my ZWave network and have NodeRed communicate with that controller?  Furthermore, if I want to run multiple NodeRed nodes, does each node require its own secondary controller?  Or is there a way to have multiple NodeRed nodes communicate together and leverage a single ZWave controller?

Thanks,

Eric

**Comment by marcus-j-davies** [comment]
Hi @benze 

> I have a HomeSeer installation with a ZWave controller. I would like to start using NodeRed for ZWave control, but can't figure out how to make both pieces work in parallel.

Only 1 software product can connect to a USB stick at a time

> but would like to attempt to control devices from NodeRed directly.

Then you must relieve Homeseer of its duty to manage Zwave (so zwave-js, can get a connection to the stick) 

With that said, if you wanted Homeseer and zwave-js to share the network, then Homeseer must become a secondary controller (assuming you have 2  zwave sticks?)

zwave-js will assume the role of primary controller, so homeseer needs to be setup as a secondary controller.

I am not a pro here, @AlCalzone  - anything important to add here?

> if I want to run multiple NodeRed nodes, does each node require its own secondary controller?

Not sure I follow here:
**node-red-contrib-zwave-js**, can only have 1 instance of the controller node per node-red install, but that controller node, will facilitate all devices, that is attached to the USB stick, you can use `zwave-device` nodes to have a node specific to each z-wave device on your network  - they will all speak to the controller node in your flow.

Remember only 1 controller node is allowed per node-red instance (but you can use as many device nodes as you want)

**Reply by AlCalzone** [reply]
> anything important to add here?

Nope, you said it all.

**Comment by benze** [comment]
> Then you must relieve Homeseer of its duty to manage Zwave (so zwave-js, can get a connection to the stick)
> 
> With that said, if you wanted Homeseer and zwave-js to share the network, then Homeseer must become a secondary controller (assuming you have 2 zwave sticks?)

Would I not be able make zwave-js attach to the secondary controller and leave HS as the primary?  Is there a reason why Zwave-js would need to be the primary?

> Remember only 1 controller node is allowed per node-red instance (but you can use as many device nodes as you want)

My question is more geared to having multiple node-red instances running on different Raspberry Pis for example.  But if I were to do that, all my zwave logic would need to be on a single Pi or alternatively have one Zwave stick/controller per Pi.  It would have been nice to be able to have NodeRed installs talk to each other such that only 1 zwave controller be needed in the cluster for all nodes to work with.

**Reply by AlCalzone** [reply]
> Is there a reason why Zwave-js would need to be the primary?

Yes, it wasn't built to be the secondary controller. I'm not sure what's necessary for that to work and there are many other features I consider to be more important right now.

**Reply by marcus-j-davies** [reply]
> It would have been nice to be able to have NodeRed installs talk to each other such that only 1 zwave controller be needed in the cluster for all nodes to work with.

I see 2 options.

Connect both (or more) NR instances together (say with MQTT) - and devise some sort of protocol between the 2.
one instance will need to be the gateway for all your devices - but each pi could have their own logic, and pass or receive from the main zwave pi.

there is a node `Event-Filter` - which can take messages that can help split incoming network events from zwave - it does not need to be attached directly to a zwave node, the same with `CMD-Factory` - you can create commands for your network - where you transport them to another pi using MQTT

OR

wait for secondary controller.
